### PR TITLE
feat!: Common Runtime in terms of Module Drivers

### PR DESCRIPTION
**Pull request by cdata**
This change proposes a re-factor of the `common-runtime` crate as series of Module Drivers. Each Module Driver constitutes a distinctive mode by which a variant of Common Module may be Prepared and Instantiated.

For example, a Module Driver may handle Preparing and Instantiating Common Functions in a Local Wasmtime-based sandbox.

What we previously called the Runtime is now unbundled; any struct can implement the `ModuleDriver` trait for a suitably constrained `ModuleDefinition`. It is expected that `ModuleDriver` will be implemented several times (once per modality that is supported by the implementor).

### CODE REVIEW

File: rust/common-macros/tests/new_type.rs [hunk 1]
```diff
@@ -6,8 +6,8 @@ fn it_derives_from_inner() {
     struct Foo(pub String);
 
     let s = String::from("foo");
-    let foo = Foo::from(s.clone());
-    assert_eq!(foo.0, s);
```
Comment (cdata) [inline rust/common-macros/tests/new_type.rs:10]: I don't know why this wasn't caught before, but clippy doesn't allow names like 'foo'.

https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_names

**Comment by cdata** [comment]
For additional context and to see what it looks like to implement a basic Module Driver, see the ongoing work in **draft** PR https://github.com/commontoolsinc/system/pull/126 where I implement a "Remote Function" driver for web.

File: rust/common-runtime/benches/runtime_bench.rs [hunk 2]
```diff
@@ -23,101 +29,136 @@ async fn init_build_server() -> Result<(Uri, JoinHandle<Result<(), BuilderError>
...
-                .into_components(Some(builder_address.clone()));
-            runtime.interpret(module, initial_io).await.unwrap()
-        };
-
-        (runtime, module_id, script_id, builder_task)
-    });
+    let (_runtime, function, function_vm, output_shape) = async_runtime
+        .block_on(async {
+            let (builder_address, _builder_task) = init_build_server().await.unwrap();
+            // let mut runtime = Runtime::new().unwrap();
```
Comment (jsantell) [inline rust/common-runtime/benches/runtime_bench.rs:63]: nit: remove

File: rust/common-runtime/src/module/definition/function_vm.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+/// A newtype over a [ModuleDefinition]; it can only be constructed for
+/// definitions whose target is [Target::CommonFunctionVm], whose affinity
+/// allows for local instantiation and whose body is [ModuleBody::SourceCode].
+#[derive(NewType)]
+#[new_type(skip(From))]
+pub struct FunctionVmDefinition(ModuleDefinition);
+
+impl FunctionVmDefinition {
+    /// Get teh [ContentType] associated with the
```
Comment (jsantell) [inline rust/common-runtime/src/module/definition/function_vm.rs:16]: nit comment

File: rust/common-runtime/src/runtime/native.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+            function_cache: Cache::new(32)?,
+            function_vm_cache: Cache::new(32)?,
+
+            vm_interpreter_cache: Cache::new(16)?,
+        })
+    }
+}
+
+#[async_trait]
```
Comment (jsantell) [inline rust/common-runtime/src/runtime/native.rs:66]: Nice consolidation here, much more legible

File: rust/common-runtime/src/serve/instantiate.rs [hunk 5]
```diff
@@ -1,81 +1,109 @@
...
-            let mut runtime = runtime.lock().await;
-            let instance_id = runtime.compile(module, initial_io).await?;
-
-            InstantiateModuleResponse {
-                module_signature: Some(module_signature),
-                instance_id: instance_id.to_string(),
-            }
+    let body = match module_reference {
+        ModuleReference::ModuleSignature(module_signature) => {
+            warn!("SIGNATURE");
```
Comment (jsantell) [inline rust/common-runtime/src/serve/instantiate.rs:37]: nit

File: rust/common-runtime/src/policy.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
+use crate::{CommonRuntimeError, InputOutput};
+use common_ifc::{Context as IfcContext, Policy};
+
+/// A type that wraps an inner value that has been validated against some [Policy].
```
Comment (jsantell) [inline rust/common-runtime/src/policy.rs:4]: What is the lifecycle of these `Validated` wrappers? To validate node input every time before executing, or once at validation time? Or maybe not yet relevant without full graphs

**Review by jsantell — APPROVED** [review]
great stuff

File: rust/common-runtime/src/policy.rs (hunk 6 above)
Comment (cdata) [reply rust/common-runtime/src/policy.rs:4]: The later. "To be defined." The only thing I wanted to do here was make it explicit that validation happened outside of the steps to invoke with some inputs.

**Pull request merged**